File: backend/src/main/java/com/ultron/backend/domain/entity/Opportunity.java
package com.ultron.backend.domain.entity;

import com.ultron.backend.domain.enums.OpportunityStage;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "opportunities")
public class Opportunity {

    @Id
    private String id;

    // Unique Opportunity ID: OPP-YYYY-MM-XXXXX
    @Indexed(unique = true)
    private String opportunityId;

    // ===== Basic Information =====
    private String opportunityName;
    private OpportunityStage stage;
    private BigDecimal amount;
    private Integer probability;  // 0-100
    private LocalDate expectedCloseDate;
    private LocalDate actualCloseDate;

    // ===== Relationships =====
    private String accountId;
    private String accountName;
    private String primaryContactId;
    private String primaryContactName;
    private String convertedFromLeadId;
    private LocalDateTime convertedDate;

    // ===== Sales Information =====
    private String type;  // New Business, Existing Business, Add-on
    private String leadSource;
    private String campaignSource;
    private String nextStep;
    private String description;

    // ===== Financial Details =====
    private BigDecimal forecastAmount;
    private String currency;
    private BigDecimal discountAmount;
    private BigDecimal totalAmount;

    // ===== Product/Service Information =====
    private List<String> products;
    private List<String> services;
    private String solutionOffered;

    // ===== Competition =====
    private List<String> competitors;
    private String competitiveAdvantage;
    private String lossReason;  // If CLOSED_LOST

    // ===== Engagement =====
    private Integer daysInStage;
    private LocalDate lastActivityDate;
    private Integer totalActivities;
    private Integer emailsSent;
    private Integer callsMade;
    private Integer meetingsHeld;

    // ===== Decision Process =====
    private String decisionMaker;
    private String decisionCriteria;
    private String budgetConfirmed;
    private LocalDate decisionTimeframe;

    // ===== Team =====
    private String ownerId;
    private String ownerName;
    private List<String> teamMembers;

    // ===== Additional Information =====
    private String deliveryStatus;
    private String paymentTerms;
    private List<String> tags;
    private String notes;

    // ===== System Fields =====
    private LocalDateTime createdAt;
    private String createdBy;
    private String createdByName;
    private LocalDateTime lastModifiedAt;
    private String lastModifiedBy;
    private String lastModifiedByName;

    // For soft delete
    private Boolean isDeleted;
    private LocalDateTime deletedAt;
    private String deletedBy;

    // ===== Stage History (optional for future) =====
    // Track stage transitions
    private LocalDateTime prospectingDate;
    private LocalDateTime qualificationDate;
    private LocalDateTime needsAnalysisDate;
    private LocalDateTime proposalDate;
    private LocalDateTime negotiationDate;
    private LocalDateTime closedDate;
}

File: frontend/types/opportunity.ts
export enum OpportunityStage {
  PROSPECTING = "PROSPECTING",
  QUALIFICATION = "QUALIFICATION",
  NEEDS_ANALYSIS = "NEEDS_ANALYSIS",
  PROPOSAL = "PROPOSAL",
  NEGOTIATION = "NEGOTIATION",
  CLOSED_WON = "CLOSED_WON",
  CLOSED_LOST = "CLOSED_LOST",
}

export interface Opportunity {
  id: string;
  opportunityId: string;

  // Basic Information
  opportunityName: string;
  stage: OpportunityStage;
  amount: number;
  probability: number;
  expectedCloseDate: string;
  actualCloseDate?: string;

  // Relationships
  accountId: string;
  accountName: string;
  primaryContactId?: string;
  primaryContactName?: string;
  convertedFromLeadId?: string;
  convertedDate?: string;

  // Sales Information
  type?: string;
  leadSource?: string;
  campaignSource?: string;
  nextStep?: string;
  description?: string;

  // Financial Details
  forecastAmount?: number;
  currency?: string;
  discountAmount?: number;
  totalAmount?: number;

  // Product/Service Information
  products?: string[];
  services?: string[];
  solutionOffered?: string;

  // Competition
  competitors?: string[];
  competitiveAdvantage?: string;
  lossReason?: string;

  // Engagement
  daysInStage: number;
  lastActivityDate?: string;
  totalActivities: number;
  emailsSent: number;
  callsMade: number;
  meetingsHeld: number;

  // Decision Process
  decisionMaker?: string;
  decisionCriteria?: string;
  budgetConfirmed?: string;
  decisionTimeframe?: string;

  // Team
  ownerId: string;
  ownerName: string;
  teamMembers?: string[];

  // Additional Information
  deliveryStatus?: string;
  paymentTerms?: string;
  tags?: string[];
  notes?: string;

  // System Fields
  createdAt: string;
  createdBy: string;
  createdByName: string;
  lastModifiedAt: string;
  lastModifiedBy: string;
  lastModifiedByName: string;

  // Stage History
  prospectingDate?: string;
  qualificationDate?: string;
  needsAnalysisDate?: string;
  proposalDate?: string;
  negotiationDate?: string;
  closedDate?: string;
}

export interface CreateOpportunityRequest {
  // Required fields
  opportunityName: string;
  stage: OpportunityStage;
  accountId: string;
  amount: number;
  probability: number;
  expectedCloseDate: string;

  // Optional fields
  primaryContactId?: string;
  type?: string;
  leadSource?: string;
  campaignSource?: string;
  nextStep?: string;
  description?: string;

  forecastAmount?: number;
  currency?: string;
  discountAmount?: number;
  totalAmount?: number;

  products?: string[];
  services?: string[];
  solutionOffered?: string;

  competitors?: string[];
  competitiveAdvantage?: string;

  decisionMaker?: string;
  decisionCriteria?: string;
  budgetConfirmed?: string;
  decisionTimeframe?: string;

  deliveryStatus?: string;
  paymentTerms?: string;
  tags?: string[];
  notes?: string;
}

export interface UpdateOpportunityRequest {
  // All fields optional
  opportunityName?: string;
  stage?: OpportunityStage;
  accountId?: string;
  primaryContactId?: string;
  amount?: number;
  probability?: number;
  expectedCloseDate?: string;
  actualCloseDate?: string;

  type?: string;
  leadSource?: string;
  campaignSource?: string;
  nextStep?: string;
  description?: string;

  forecastAmount?: number;
  currency?: string;
  discountAmount?: number;
  totalAmount?: number;

  products?: string[];
  services?: string[];
  solutionOffered?: string;

  competitors?: string[];
  competitiveAdvantage?: string;
  lossReason?: string;

  decisionMaker?: string;
  decisionCriteria?: string;
  budgetConfirmed?: string;
  decisionTimeframe?: string;

  deliveryStatus?: string;
  paymentTerms?: string;
  tags?: string[];
  notes?: string;
}

export interface OpportunityStatistics {
  // Total counts
  totalOpportunities: number;
  openOpportunities: number;
  closedOpportunities: number;

  // By stage
  prospectingCount: number;
  qualificationCount: number;
  needsAnalysisCount: number;
  proposalCount: number;
  negotiationCount: number;
  wonCount: number;
  lostCount: number;

  // Financial metrics
  totalValue: number;
  wonValue: number;
  lostValue: number;
  pipelineValue: number;
  weightedValue: number;

  // Performance metrics
  winRate: number;
  averageDealSize: number;
  averageCloseDays: number;
}
